src/qutip_tensornetwork/core/data/network.py

```diff
@@ -0,0 +1,576 @@
+    `tensornetwork.Edges`. Nodes represent n-dimensional arrays whose dimensions
+    are connected with edges to other arrays. An edge represents the
+    contraction that needs to be carried out between two nodes. The contraction
+    of nodes is done in a `lazy` way, which means that is only carried out
+    when necessary or explicitly required with the contract method.
+
+    Edges can be either dangling (one of its ends is not connected to any node)
+    or fully connected. To represents arbitrary quantum objects, we
+    employ the dangling edges `out_edges` and `in_edges`. Considered as a matrix, the
+    dangling `out_edges` and `in_edges` represent the rows and columns of the
+    matrix respectively.
+
+    Notes:
+    -----
+    Most of the operations and logic of this class has been derived from
+    `tensornetwork.QuOperator`. However, this class is not compatible with
+    `QuOperator` but we provide a method, `as_quoperator()` that returns a view
+    of this class as a `QuOperator`.
+    """
+
+    def __init__(
+        self,
+        out_edges: Sequence[Edge],
+        in_edges: Sequence[Edge],
+        nodes: Optional[Collection[AbstractNode]] = None,
+        copy: Optional[bool] = True) -> None:
+        """Creates a new `Netowork` from a tensor network.
+        This encapsulates an existing tensor network, interpreting it as a linear
+        operator.
+
+        The network is checked for consistency: All dangling edges must either be
+        in `out_edges`, `in_edges`, or `ignore_edges`.
+
+        Args:
+            out_edges:
+                The edges of the network to be used as the output edges.
+            in_edges:
+                The edges of the network to be used as the input edges.
+            nodes:
+                Nodes of the network. If none provided, the nodes are obtained
+                by finding all the nodes that belong to the networks form by
+                in_edges and out_edges.
+        """
+        if (len(in_edges) == 0 and len(out_edges) == 0
+            and (nodes is None or len(nodes)==0)):
+            raise ValueError("Since no edges were provided, it was not possible"
+                             "to infer which nodes belong to the network."
+                             "You may want to include a scalar node to represent"
+                             "a matrix with shape")
+
+        # I need to check that edges are unique
+        self.out_edges = list(out_edges)
+        self.in_edges = list(in_edges)
+
+        # Unlike in QuOperator we will keep track of the nodes instead of
+        # dynamically searching for them when necessary. This is because
+        # searching all nodes in a large graph can be quite expensive while
+        # keeping track of them with network operations is straightforward.
+        self.nodes = set(nodes) if nodes else tn.reachable(self.in_edges
+                                                         + self.out_edges)
```

Comment:
I am not sure how to store nodes. They need to be unique which is why in TensorNetwork they mostrly use sets. However, this means that adding a new node will involve checking all of them to ensure the new node is unique. In most cases this is unnecessary. For instance, in matmul we copy both input networks so we now that the nodes are unique. I am thinking to change this into a list. Any thought @hodgestar, @nwlambert ?

Reply:
I think a set is fine for now. Checking "x in some_set" is very fast in Python (similar to "x in some_dict").